core/src/Cuda/Kokkos_Cuda.hpp

```diff
@@ -171,7 +171,18 @@ class Cuda {
 
   KOKKOS_DEPRECATED Cuda(cudaStream_t stream, bool manage_stream);
 
+  ~Cuda() {
+    // In case of multi-gpu, we must free scratch before the
+    // internal space is deleted so that we can pass a valid execution
+    // space to kokkos_free().
+    free_scratch();
+  }
```

Comment:
We get a segfault if we attempt to create a non-default Cuda object inside `CudaInternal::finalize()`, so we cannot free the scratch space on the correct device. 

Not sure exactly why we segfault, but I can look into that more if we don't like this solution.